File: packages/cem-elements/src/lib/disposition.ts
/**
 * BR-VC-9 unknown-optional-feature disposition — the FF-4 decision-core.
 *
 * When a governed contract payload carries an **unknown optional feature**
 * (e.g. an additive field from a higher MINOR than this build understands),
 * the engine selects one of three dispositions, chosen by the effective
 * **run mode** and the **contract's class**:
 *
 *   - **Application** (default) — *per-contract*: tolerant on presentation
 *     contracts (templates, tokens), strict (reject) on data/security
 *     contracts (snapshot/`datadom`, privacy export, edge render-state).
 *   - **Build / SSR** — *strict*: reject every optional unknown across all
 *     contracts, so build artifacts / server-rendered output never silently
 *     drop content.
 *   - **Development** — *tolerant*: degrade (accept + continue, surfacing
 *     diagnostics) across all contracts, so work-in-progress is not
 *     hard-stopped.
 *
 * The **must-understand reject** of BR-VC-8 is a separate axis and holds in
 * **every** mode: a feature flagged must-understand always rejects,
 * regardless of the optional-feature disposition.
 *
 * This module is the pure decision-core only (BR-VC-9 §6.5 / AC-P-6.7).
 * Wiring it to the actual contract ingest paths (snapshot / edge-render-state
 * read-back) and threading the run mode from the host/build config are a
 * follow-up slice; this core carries no I/O and no global state so it is
 * deterministic and exhaustively unit-testable. It backs fitness function
 * FF-4 (`tools/fitness/fitness-gates.json`).
 */

/** Run mode selecting the unknown-optional-feature disposition (BR-VC-9). */
export type RunMode = 'application' | 'build-ssr' | 'development';

/**
 * Disposition applied to a single unknown OPTIONAL feature.
 * - `reject` — strict: the unknown feature is an error; the host must not
 *   silently proceed.
 * - `degrade` — tolerant: accept the payload without the unknown feature's
 *   semantics (treat as unvalidated/foreign) and continue, surfacing a
 *   diagnostic.
 * - `ignore` — tolerant: drop the unknown feature and continue, surfacing a
 *   diagnostic (a report event).
 */
export type Disposition = 'reject' | 'degrade' | 'ignore';

/**
 * BR-VC-9 contract classes. Presentation contracts tolerate unknown optional
 * features in an application run; data/security contracts do not.
 */
export type ContractClass = 'presentation' | 'data-security';

/**
 * Governed contracts that can carry versioned optional features (mirrors
 * `tools/fitness/governed-contracts.json` plus the BR-VC-9-named
 * `privacy-export`, which has no SemVer locator yet).
 */
export type GovernedContractId =
    | 'data-snapshot'
    | 'edge-render-state'
    | 'privacy-export'
    | 'template-authoring-cem-ml'
    | 'token-outputs'
    | 'patch-transport';

/**
 * Static BR-VC-9 classification of each governed contract.
 *
 * Data/security (strict in an application run): the snapshot/`datadom`
 * contract, privacy export, and edge render-state — these cross trust /
 * serialization boundaries, so an unrecognized optional feature must not be
 * silently honored or dropped. Presentation (tolerant): templates and
 * tokens. `patch-transport` is the render-plan/patch stream — rendering
 * output, classed presentation; it is not named explicitly by BR-VC-9, so
 * the conservative reviewer note lives here.
 */
const CONTRACT_CLASS: Readonly<Record<GovernedContractId, ContractClass>> = {
    'data-snapshot': 'data-security',
    'edge-render-state': 'data-security',
    'privacy-export': 'data-security',
    'template-authoring-cem-ml': 'presentation',
    'token-outputs': 'presentation',
    'patch-transport': 'presentation',
};

/** BR-VC-9 class of a governed contract. */
export function classifyContract(contract: GovernedContractId): ContractClass {
    return CONTRACT_CLASS[contract];
}

/** A single, auditable disposition decision (BR-VC-9 "record the active mode"). */
export interface DispositionDecision {
    mode: RunMode;
    contract: GovernedContractId;
    contractClass: ContractClass;
    disposition: Disposition;
    /** True when the unknown feature is rejected (strict outcome). */
    strict: boolean;
    /** True when the reject is forced by BR-VC-8 must-understand, not the optional-feature policy. */
    mustUnderstand: boolean;
    /** Human-readable rationale citing the governing rule. */
    rationale: string;
}

/** The tolerant disposition chosen per run mode (ignore/degrade per BR-VC-9). */
const TOLERANT_BY_MODE: Readonly<Record<RunMode, Exclude<Disposition, 'reject'>>> = {
    // An application run keeps presentation content but treats the unknown
    // optional feature as unvalidated — degrade rather than drop.
    application: 'degrade',
    // build-ssr is never tolerant (handled before this table is consulted).
    'build-ssr': 'degrade',
    // Development preserves work-in-progress: degrade + surface diagnostics.
    development: 'degrade',
};

/**
 * Decide the BR-VC-9 disposition for an unknown optional feature on
 * `contract` under `mode`. Pure and total over the typed domain.
 *
 * `mustUnderstand` (BR-VC-8): when the unknown feature is flagged
 * must-understand it rejects in every mode, ahead of the optional-feature
 * policy.
 */
export function decideDisposition(
    mode: RunMode,
    contract: GovernedContractId,
    opts: { mustUnderstand?: boolean } = {},
): DispositionDecision {
    const contractClass = classifyContract(contract);
    const mustUnderstand = opts.mustUnderstand === true;

    if (mustUnderstand) {
        return {
            mode,
            contract,
            contractClass,
            disposition: 'reject',
            strict: true,
            mustUnderstand: true,
            rationale: 'BR-VC-8: a must-understand feature rejects in every run mode.',
        };
    }

    // BR-VC-9 mode rules.
    let strict: boolean;
    let rationale: string;
    switch (mode) {
        case 'build-ssr':
            strict = true;
            rationale =
                'BR-VC-9: build/SSR rejects every optional unknown across all contracts.';
            break;
        case 'development':
            strict = false;
            rationale =
                'BR-VC-9: development tolerates (degrades) optional unknowns across all contracts, surfacing diagnostics.';
            break;
        case 'application':
        default:
            strict = contractClass === 'data-security';
            rationale = strict
                ? 'BR-VC-9: an application run rejects optional unknowns on data/security contracts.'
                : 'BR-VC-9: an application run tolerates (degrades) optional unknowns on presentation contracts.';
            break;
    }

    return {
        mode,
        contract,
        contractClass,
        disposition: strict ? 'reject' : TOLERANT_BY_MODE[mode],
        strict,
        mustUnderstand: false,
        rationale,
    };
}

File: packages/cem_ml/src/schema/disposition.rs
//! AC-P-6.7 / AC-P-V-6 unknown-namespace disposition — decision-core.
//!
//! When a region's namespace resolves to **no metadata, no explicit schema,
//! and no rule**, the effective scope policy MUST select one defined behavior:
//!
//! - [`Disposition::Reject`] — a hard error; the region is not accepted.
//! - [`Disposition::Allow`] — accept as unvalidated foreign content.
//! - [`Disposition::Ignore`] — drop the region with a report event.
//!
//! The **default** is mode-selected (the BR-VC-9 run-mode disposition): an
//! application run rejects unknown data/security namespaces and allows unknown
//! presentation namespaces; build/SSR rejects all; development allows all. A
//! scope-policy rule MAY override within the mode, and the outcome MUST be
//! deterministic. `allow` and `ignore` are non-execution modes unless a
//! separate handler is explicitly selected.
//!
//! This module is the pure decision-core only. Detecting an unresolved-namespace
//! region and applying the outcome in the schema machine — plus threading the
//! run mode through the machine — is a follow-up slice. Pure, total over the
//! typed domain, and no I/O, so it is exhaustively unit-testable. It is the
//! parser-side counterpart to the cem-elements contract disposition that backs
//! fitness function FF-4; this side is the literal AC-P-V-6 verifier (tracked
//! in `docs/todo.md`).

/// Run mode selecting the default disposition for an unresolved-namespace region.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RunMode {
    Application,
    BuildSsr,
    Development,
}

/// Defined behavior for a region whose namespace resolves to nothing.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Disposition {
    /// Hard error — the region is not accepted.
    Reject,
    /// Accept as unvalidated foreign content.
    Allow,
    /// Drop the region with a report event.
    Ignore,
}

/// Class hint for an unknown namespace, when a scope-policy rule (or host
/// metadata) provides one. Absent a hint the namespace is unclassified.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum NamespaceClass {
    Presentation,
    DataSecurity,
}

/// Where the resolved disposition came from — recorded for auditability
/// (BR-VC-9 "record the active mode … so the disposition is auditable").
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DispositionSource {
    RunModeDefault,
    ScopePolicyOverride,
}

/// A single, auditable disposition decision for an unresolved-namespace region.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct DispositionDecision {
    pub mode: RunMode,
    pub class: Option<NamespaceClass>,
    pub disposition: Disposition,
    pub source: DispositionSource,
}

/// The mode-selected default disposition per AC-P-6.7 / BR-VC-9.
///
/// Build/SSR rejects every unknown namespace; development allows every unknown
/// namespace; an application run rejects unknown data/security namespaces and
/// allows unknown presentation namespaces. An **unclassified** unknown namespace
/// in an application run is rejected conservatively (treated as potentially
/// data/security until a scope-policy rule classifies it otherwise).
pub fn default_disposition(mode: RunMode, class: Option<NamespaceClass>) -> Disposition {
    match mode {
        RunMode::BuildSsr => Disposition::Reject,
        RunMode::Development => Disposition::Allow,
        RunMode::Application => match class {
            Some(NamespaceClass::Presentation) => Disposition::Allow,
            Some(NamespaceClass::DataSecurity) => Disposition::Reject,
            None => Disposition::Reject,
        },
    }
}

/// Resolve the effective disposition. A scope-policy `policy_override` wins over
/// the mode default (AC-P-6.7 "Scope policy MAY override within the mode"); the
/// outcome is deterministic for a given `(mode, class, policy_override)`.
pub fn resolve_disposition(
    mode: RunMode,
    class: Option<NamespaceClass>,
    policy_override: Option<Disposition>,
) -> DispositionDecision {
    match policy_override {
        Some(disposition) => DispositionDecision {
            mode,
            class,
            disposition,
            source: DispositionSource::ScopePolicyOverride,
        },
        None => DispositionDecision {
            mode,
            class,
            disposition: default_disposition(mode, class),
            source: DispositionSource::RunModeDefault,
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const CLASSES: [Option<NamespaceClass>; 3] = [
        None,
        Some(NamespaceClass::Presentation),
        Some(NamespaceClass::DataSecurity),
    ];

    #[test]
    fn build_ssr_rejects_every_class() {
        for class in CLASSES {
            assert_eq!(default_disposition(RunMode::BuildSsr, class), Disposition::Reject);
        }
    }

    #[test]
    fn development_allows_every_class() {
        for class in CLASSES {
            assert_eq!(default_disposition(RunMode::Development, class), Disposition::Allow);
        }
    }

    #[test]
    fn application_is_per_class() {
        assert_eq!(
            default_disposition(RunMode::Application, Some(NamespaceClass::Presentation)),
            Disposition::Allow
        );
        assert_eq!(
            default_disposition(RunMode::Application, Some(NamespaceClass::DataSecurity)),
            Disposition::Reject
        );
        // Unclassified → conservative reject.
        assert_eq!(default_disposition(RunMode::Application, None), Disposition::Reject);
    }

    #[test]
    fn scope_policy_override_wins_in_every_mode() {
        let modes = [RunMode::Application, RunMode::BuildSsr, RunMode::Development];
        let overrides = [Disposition::Reject, Disposition::Allow, Disposition::Ignore];
        for mode in modes {
            for class in CLASSES {
                for over in overrides {
                    let decision = resolve_disposition(mode, class, Some(over));
                    assert_eq!(decision.disposition, over);
                    assert_eq!(decision.source, DispositionSource::ScopePolicyOverride);
                }
            }
        }
    }

    #[test]
    fn no_override_uses_the_mode_default() {
        let decision = resolve_disposition(RunMode::Application, Some(NamespaceClass::Presentation), None);
        assert_eq!(decision.disposition, Disposition::Allow);
        assert_eq!(decision.source, DispositionSource::RunModeDefault);

        let rejected = resolve_disposition(RunMode::BuildSsr, Some(NamespaceClass::Presentation), None);
        assert_eq!(rejected.disposition, Disposition::Reject);
        assert_eq!(rejected.source, DispositionSource::RunModeDefault);
    }

    #[test]
    fn decision_record_echoes_inputs() {
        let decision = resolve_disposition(RunMode::Application, Some(NamespaceClass::DataSecurity), None);
        assert_eq!(decision.mode, RunMode::Application);
        assert_eq!(decision.class, Some(NamespaceClass::DataSecurity));
    }

    #[test]
    fn ignore_is_only_reachable_via_an_explicit_override() {
        // The mode defaults never select Ignore; it is an opt-in scope-policy choice.
        for mode in [RunMode::Application, RunMode::BuildSsr, RunMode::Development] {
            for class in CLASSES {
                assert_ne!(default_disposition(mode, class), Disposition::Ignore);
            }
        }
        assert_eq!(
            resolve_disposition(RunMode::Application, None, Some(Disposition::Ignore)).disposition,
            Disposition::Ignore
        );
    }
}
